src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+    //  .div(self.tokenPrecisionMultipliers[tokenIndex]).sub(dy)
+    const dySwapFee =
+      (this._xp(state)[tokenIndex] - newY) /
+        state.tokenPrecisionMultipliers[tokenIndex] -
+      dy;
+
+    // dy = dy.mul(FEE_DENOMINATOR.sub(calculateCurrentWithdrawFee(self, account)))
+    //  .div(FEE_DENOMINATOR);
+    dy =
+      (dy * (this.FEE_DENOMINATOR - this._calculateCurrentWithdrawFee(state))) /
+      this.FEE_DENOMINATOR;
+
+    return { dy, dyFee: dySwapFee };
+  }
+
+  swapUnderlying(state: PoolState) {}
+
+  protected _getNumTokens(state: PoolState) {
+    return biginterify(state.tokenPrecisionMultipliers.length);
+  }
+
+  protected _calculateWithdrawOneTokenDY(
+    state: PoolState,
+    tokenIndex: number,
+    tokenAmount: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const numTokens = this._getNumTokens(state);
+    const xp = this._xp(state);
+    const v = {
+      d0: ZERO,
+      d1: ZERO,
+      newY: ZERO,
+      feePerToken: ZERO,
+      preciseA: ZERO,
+    };
+    v.preciseA = this._getAPrecise(state, blockTimestamp);
+    const d0 = this._getD(state, xp, v.preciseA);
+    if (d0 === null) return null;
+    else v.d0 = d0;
+
+    // v.d1 = v.d0.sub(tokenAmount.mul(v.d0).div(self.lpToken.totalSupply()));
+    v.d1 = v.d0 - (tokenAmount * v.d0) / state.lpToken_supply;
+    const newY = this._getYD(state, v.preciseA, tokenIndex, xp, v.d1);
+    if (newY === null) return null;
+    else v.newY = newY;
+
+    const xpReduced: bigint[] = [];
+
+    v.feePerToken = this._feePerToken(state);
+    for (let i = 0; i < numTokens; i++) {
+      const xpi = xp[i];
+      // if i == tokenIndex, dxExpected = xp[i] * d1 / d0 - newY
+      // else dxExpected = xp[i] - (xp[i] * d1 / d0)
+      // xpReduced[i] -= dxExpected * fee / FEE_DENOMINATOR
+
+      const dxExpected =
+        i === tokenIndex
+          ? (xpi * v.d1) / v.d0 - v.newY
+          : xpi - (xpi * v.d1) / v.d0;
+      xpReduced[i] -= (dxExpected * v.feePerToken) / this.FEE_DENOMINATOR;
+    }
+
+    const yd = this._getYD(state, v.preciseA, tokenIndex, xpReduced, v.d1);
+    if (yd === null) return null;
+    // uint256 dy = xpReduced[tokenIndex].sub(getYD(v.preciseA, tokenIndex, xpReduced, v.d1));
+    let dy = xpReduced[tokenIndex] - yd;
+
+    // dy = dy.sub(1).div(self.tokenPrecisionMultipliers[tokenIndex]);
+    dy = (dy - ONE) / state.tokenPrecisionMultipliers[tokenIndex];
+
+    return { dy, newY: v.newY };
+  }
+
+  protected _feePerToken(state: PoolState) {
+    const numTokens = this._getNumTokens(state);
+    // self.swapFee.mul(self.pooledTokens.length).div(self.pooledTokens.length.sub(1).mul(4));
+    return (state.swapFee * numTokens) / ((numTokens - ONE) * biginterify(4));
+  }
+
+  protected _calculateCurrentWithdrawFee(state: PoolState) {
+    // It is not correct. We should calculate user withdrawFeeMultiplier by
+    // the time passed since the liquidity was added
+    return state.defaultWithdrawFee ? state.defaultWithdrawFee : ZERO;
+  }
```

Comment:
how is it handled when the withdraw event is released?

Reply:
Actually, we don't use this as its not fully correct. I just invalidate the state, so it can be updated by RPC call